gems/sorbet-runtime/lib/types/private/methods/_methods.rb

```diff
@@ -121,6 +135,26 @@ def self._on_method_added(hook_mod, method_name, is_singleton_method: false)
     mod = is_singleton_method ? hook_mod.singleton_class : hook_mod
     original_method = mod.instance_method(method_name)
 
+    if !current_declaration.nil? && current_declaration.final?
+      if !mod.is_a?(Class)
+        raise "`#{mod.name}` is not a class and its method `#{method_name}` cannot be declared as final"
+      end
```

Comment:
You'll have to check with Dmitry but this might be too restrictive. We might want `final` methods in modules maybe?

Reply:
OK to get @DarkDimius's opinion on this.

Hmm. My initial thought was that since you can't have a Module that's not a Class as a superclass, it doesn't make sense to mark a Module as final. But then I suppose a final method on a Module would make sense, since I suppose that means if you include that Module, you can't define your own version of the final method. And so then maybe a final Module makes sense too because it could just mean "every method in this Module is final".

That reminds me, I should make it so that if you mark a class final, every method in the class gets marked as final as well. Because we want to defend against define_method and things like that.